python/plugins/processing/algs/qgis/TilesXYZ.py

```diff
@@ -0,0 +1,366 @@
+    EXTENT = 'EXTENT'
+    ZOOM_MIN = 'ZOOM_MIN'
+    ZOOM_MAX = 'ZOOM_MAX'
+    TILE_FORMAT = 'TILE_FORMAT'
+    DPI = 'DPI'
+    OUTPUT_FORMAT = 'OUTPUT_FORMAT'
+    OUTPUT_DIRECTORY = 'OUTPUT_DIRECTORY'
+
+    def group(self):
+        return self.tr('Raster tools')
+
+    def groupId(self):
+        return 'rastertools'
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterString(self.NAME, self.tr('Name'), defaultValue='Test'))
+        self.addParameter(QgsProcessingParameterFolderDestination(self.OUTPUT_DIRECTORY, self.tr('Output directory')))
+
+        self.addParameter(QgsProcessingParameterExtent(self.EXTENT, self.tr('Extent')))
+        self.addParameter(QgsProcessingParameterNumber(self.ZOOM_MIN,
+                                                       self.tr('Minimum zoom'),
+                                                       defaultValue=12))
+        self.addParameter(QgsProcessingParameterNumber(self.ZOOM_MAX,
+                                                       self.tr('Maximum zoom'),
+                                                       defaultValue=12))
+        self.addParameter(QgsProcessingParameterNumber(self.DPI, self.tr('DPI'), defaultValue=96))
+
+        self.formats = ['PNG', 'JPG']
+        self.addParameter(QgsProcessingParameterEnum(self.TILE_FORMAT,
+                                                     self.tr('Tile format'),
+                                                     self.formats,
+                                                     defaultValue=0))
+        self.outputs = ['Directory', 'MBTiles']
+        self.addParameter(QgsProcessingParameterEnum(self.OUTPUT_FORMAT,
+                                                     self.tr('Output format'),
+                                                     self.outputs,
+                                                     defaultValue=0))
+
+    def name(self):
+        return 'tilesxyz'
+
+    def displayName(self):
+        return self.tr('Generate XYZ tiles')
+
+    def processAlgorithm(self, parameters, context, feedback):
+        feedback.setProgress(1)
+
+        name = self.parameterAsString(parameters, self.NAME, context)
+        output_dir = self.parameterAsString(parameters, self.OUTPUT_DIRECTORY, context)
+        extent = self.parameterAsExtent(parameters, self.EXTENT, context)
+        min_zoom = self.parameterAsInt(parameters, self.ZOOM_MIN, context)
+        max_zoom = self.parameterAsInt(parameters, self.ZOOM_MAX, context)
+        dpi = self.parameterAsInt(parameters, self.DPI, context)
+        tile_format = self.formats[self.parameterAsEnum(parameters, self.TILE_FORMAT, context)]
+        output_format = self.outputs[self.parameterAsEnum(parameters, self.OUTPUT_FORMAT, context)]
+        tile_width = 256
+        tile_height = 256
+
+        project = QgsProject.instance()
+        visible_layers = [item.layer() for item in project.layerTreeRoot().findLayers() if item.isVisible()]
```

Comment:
This isn't safe to run in a background thread -- it needs to be moved to a prepareAlgorithm step and run in the main thread.

Reply:
I have moved accessing of project's layers into the prepareAlgorithm. If some other parts will need to be moved as well, I think @wonder-sk will take care of it.